sim/src/service/engineService.cpp

```diff
@@ -5,17 +5,163 @@
+    ServerContext* context, const bento::protos::StepSimulationReq* request,
+    bento::protos::StepSimulationResp* response) {
+    if (!sims.contains(request->name())) {
+        return Status(grpc::NOT_FOUND,
+                      "Could not find simulation with that name.");
+    }
+
+    auto& sim = sims.at(request->name());
+
+    // TODO: Run built-in systems as well
+    auto& indexStore = sim->indexStore;
+    auto& compStore = sim->compStore;
+    auto& simDef = sim->simDef;
+
+    // Lock the simulation
+    if (!sim->locked) {
+        sim->locked = true;
+        interpreter::runGraph(compStore, indexStore, sim->simDef.init_graph());
+    }
+
+    for (size_t i = 0; i < simDef.systems_size(); i++) {
+        const auto& graph = simDef.systems(i).graph();
+        interpreter::runGraph(compStore, indexStore, graph);
+    }
+
+    return Status::OK;
+}
+
+Status EngineServiceImpl::GetAttribute(
+    ServerContext* context, const bento::protos::GetAttributeReq* request,
+    bento::protos::GetAttributeResp* response) {
+    if (!sims.contains(request->sim_name())) {
+        return Status(grpc::NOT_FOUND,
+                      "Could not find simulation with that name.");
+    }
+
+    auto& indexStore = sims.at(request->sim_name())->indexStore;
+    auto& compStore = sims.at(request->sim_name())->compStore;
+    auto& simDef = sims.at(request->sim_name())->simDef;
+
+    // Use interpreter's operations to find the attribute for consistency
+    // Create the retrieve node
+    auto retrieveNode = bento::protos::Node_Retrieve();
+    retrieveNode.mutable_retrieve_attr()->CopyFrom(request->attribute());
+
+    auto val = interpreter::retrieveOp(compStore, indexStore, retrieveNode);
+    // CopyFrom does not accept rvalue references, so there is no speed-up from
+    // moving
+    response->mutable_value()->CopyFrom(val);
+
+    return Status::OK;
+}
+
+Status EngineServiceImpl::SetAttribute(
+    ServerContext* context, const bento::protos::SetAttributeReq* request,
+    bento::protos::SetAttributeResp* response) {
+    if (!sims.contains(request->sim_name())) {
+        return Status(grpc::NOT_FOUND,
+                      "Could not find simulation with that name.");
+    }
+
+    auto& indexStore = sims.at(request->sim_name())->indexStore;
+    auto& compStore = sims.at(request->sim_name())->compStore;
+    auto& simDef = sims.at(request->sim_name())->simDef;
+
+    // Use interpreter's node to find the attribute for consistency
+    // Create a const node holding the value
+    auto valConstNode = bento::protos::Node();
+    valConstNode.mutable_const_op()->mutable_held_value()->CopyFrom(
+        request->value());
+    // Create the retrieve node
+    auto mutateNode = bento::protos::Node_Mutate();
+    mutateNode.mutable_mutate_attr()->CopyFrom(request->attribute());
+    mutateNode.mutable_to_node()->CopyFrom(valConstNode);
+
+    interpreter::mutateOp(compStore, indexStore, mutateNode);
+
+    return Status::OK;
+}
```

Comment:
Missing  `grpc::NOT_FOUND` status code for when the user specifies a `AttributeRef` that does not refer to any attributes that have been registered.

Reply:
I have added more exception handling in 968ac35.

However, for `SetAttribute` I will need more exception handling on the operation level if I want to determine whether or not the problem is in evaluating a node's value or with retrieving a attribute. Hence, I have used `grpc::UNKNOWN` for that.

Reply:
I think those are more `500` Internal Server Errors.
I think we should be using `grpc::INTERNAL` for those errors.